### DISCUSSION TITLE
[Feature] Two-way Sync

### DISCUSSION THREAD

**Discussion by partnerinflight** [immich-app/immich#4341] [Feature Request]
### The feature

This has been requested in various forms, but wanted to generalize a bit. I'd love to see a way to indicate the type of sync Immich performs. Specifically:

* Default one-way sync: assets sync from Photos to Immich only. 
* One-way smart sync: assets deleted from Immich aren't re-synced from cloud. 
* Two-way sync: assets deleted in Photos are deleted from Immich. Assets deleted from Immich are deleted in Photos.

I would at least ask for the one-way smart sync; without that, Immich's usability as a primary photo browsing destination is limited, as I frequently find things synced to Immich I actually want to delete. If they're just going to resync again, that would be useless.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by dgsardina** [comment]
I think there are two main cases here. Photos that want be deleted in both server and mobile due to being discarded, and photos that want to be kept in the server but deleted in the mobile to save space in the latter. And both are required in a normal usage (we all take bad pics and have limited storage in the mobile devices).

**Reply by mckennajones** [reply]
I agree with this. My main use case is that I still want to use the default iPhone Photos app to manage my photos, so it would be great if photos deleted there are also deleted from Immich. Although this feels like maybe it shouldn't be the default behavior, and instead toggled on/off via a setting.

**Comment by Kieranwest** [comment]
I think it would also be great for a cloud only option. Have files deleted locally once they're synced to Immich. Similar to how ES File Explorer works when you import photos. It prompts for the deletion of the synced files.

**Comment by andrewdcaudle** [comment]
I'm specifically looking for two-way sync.

**Comment by Havouza15** [comment]
I sometime want to keep photos on the phone even if I delete them on Immich so that possibility must exist like now.

**Comment by markelarrasate** [comment]
Hi, I add a new proposal for a new option for this sync feature.

### "Free up space" option

**Description**: Add a "Free up space" option in the Immich mobile app that allows users to delete photos and videos from their device while retaining them in Immich storage.

**Details**:
- **Accessibility**: Include in the main menu or settings.
- **Functionality**: Users select items to delete from the device, keeping them in Immich cloud storage.
- **Confirmation**: Show a warning to confirm the action, clarifying that items will only be removed from the device.
- **Notifications**: Inform users when items are successfully deleted from the device.

**Benefits**:
- **Space optimization**: Frees up device storage.
- **Data security**: Ensures items remain safe in Immich.
- **Ease of use**: Simplifies storage management on mobile devices.

**Reply by bo0tzz** [reply]
#165

**Comment by Lyessi78** [comment]
I think we can add at this scope adding and modifying. 

Two way sync : add in the cloud if photo is added in album from mobile and download in the mobile if photo is added from web or another phone
One way sync (mobile -> cloud): add in the cloud if photo is added in album from mobile
One way sync (sync -> mobile): add in the mobile all the new photo that are added in the cloud

The photo is download in the mobile in the same folder that the album concerned and must works with shared album.

**Comment by aloxyz** [comment]
I second this. Deleting everything twice is painful

**Comment by H4usi** [comment]
I agree with this. There needs to be a way to manage the pictures more easily. 

I like to manage my pictures in the iCloud or on the Immich website. But this means that they are either not deleted on the mobile phone/iCloud or on the Immich server and I have to do everything twice. Which is just no fun. 

Maybe we can meet in the middle, if a two-way-sync is not planned, maybe also for technical reasons, it might be possible to add a filter to the gallery of Immich on the mobile phone. 

One that only shows you the pictures that are on your mobile phone but not on Immich and one that shows you pictures that are only on Immich but not on your mobile phone. This would allow you to quickly compensate for the asynchrony.

**Reply by timaschew** [reply]
I agree, deleting twice isn't fun.
The reason for this is, that sometimes it's easier to manage the fotos (including deletion) from the browser than from the phone, right?

I have another idea, which is probably much easer to implement. Let me try to explain:

1. User is deleting lot of photos via the immich web app (assets are moved to trash only, cloud trash).
2. User is swtiching to the mobile app and enabling a new filter type (show assets in cloud-trash)
3. User can select some or all assets which are in the cloud trash and decide then to click a button to remove them from the local device)

The whole mechanism could be user also to filter asses which are already in the cloud. The user could then free up space from the device without moving them to the cloud trash.

**Comment by dffvb** [comment]
I really need this…. Flying by plane, going through my 30k pictures, deleting them and have it synced

**Reply by panzeyu2013** [reply]
I am doing the same!

**Comment by panzeyu2013** [comment]
I think it is an easier but also functional way to automatically create two folders. One that users can select important images that stores permanently. Another one filters out those stored on the cloud but not on the local device to help users quickly identify remove these photos.

**Comment by Demian98** [comment]
I'm currently using nextcloud in combination with the FolderSync Android App to two-way sync my photos.

On the phone this is working very well, especially with the native gallery, but the Nextcloud photos App is very crappy. 

Can't wait to switch all of this to Immich😊 But two-way sync is a must-have feature for me, hope to see it soon 👍

**Comment by XDelta778** [comment]
Is there anyone working on this feature. I think it wouild be very useful for us all !

(Just installed immich within hours thinking it was already up.....)

**Comment by majorsl** [comment]
I came here looking for this feature too. Having to manage individual photos across multiple devices/apps seems like a recipe to lose things.  If I don't want it, I don't want it - anywhere.

Even if trashed on the phone, it could be in Immich's Trash which offers a recovery path for an "oops."

Agreed - is this being worked on?

**Comment by Demian98** [comment]
I developed a solution (https://github.com/immich-app/immich/discussions/20522) to get this working. Would love to get your feedback to this

**Comment by PargsZ** [comment]
Any official feedback if two-way sync will be implemented somehow? 
This really would be great.

**Comment by jimmyli56** [comment]
I do agree this is great to have, as I am currently struggling to setup/find an two way sync system at the moment, so currently resorting to just a one way backup using the mobile app/photosync, waiting out for a two way system. I would much prefer to keep Apple Photos as a primary means of accessing my photos and using Immich as a local, self hosted backup/archive and have both sync to deletions. Hoping this is still being considered as an upcoming feature!

**Reply by KiLLeRRaT** [reply]
This is my use case too.  I want to have Immich perform deletions as I delete things from Photos on the iPhone.  If we can do that without using other software it'd be great :)


## Slightly off topic
Sorry for the off topic, but it'd be good to get the take of folks that also want 2 way sync, and see what workflows they have, or workflows they'd like.

I'm still trying to figure out a solid workflow.  If I can somehow stop my iCloud storage subscription, that'd be ideal.  

- Take photo on iPhone
- Edit
    - Photos app
    - Lightroom mobile (I typically export to JPG) after editing 
    - VSCO
- Manage photos by moving to albums in Photos app
- Have all these photos/edits and albums sync'd to Immich
- Can delete photos in Immich, or in Photos app and it's mirrored on both sides.  When at home, I might sort out photos on the computer using Immich, so it should sync deletions back to iPhone.  When on the run, I might sort out AND edit photos using Photos app, and that should sync back to Immich so I can pick up things later.

Any recommendations or suggestions?

**Reply by jimmyli56** [reply]
At the moment, it does seem that on the Immich mobile app on iOS, I can: 

- Delete a photo that's stored both on device and immich, and the app will first delete the immich copy, and prompt me whether or not to delete the photo off the iPhone as well. (Which is better than what I've previously imagined and feels very much like Google Photos sync that I am used to). 

This is somewhat workable with me, and I am fine with that. I do wish I could: 

- Handle bulk culling on another platform (i.e., web), and have the app sync up deletions so I can also remove them from my iPhone/iCloud, as when I delete a photo on my computer, it disappears from the mobile app.

Now I just have to figure out how to store my RAW (and edited JPEG) photos (as well as previewing, file handling, file syncing, workflow, all that fun stuff) from Lightroom/other apps, and I think I'll be a happy camper.

**Comment by rrakso** [comment]
Synology Photos has solution for it -> https://github.com/immich-app/immich/discussions/11344#discussioncomment-15983651

**Reply by jimmyli56** [reply]
Bingo, that's exactly what I'm looking for.

**Comment by djafrika** [comment]
Backup is not sync... Immich has only backup.

Synchronization is having identical folders and files on 2 or more devices.
This means that there are 2 or more identical copies on phone, server, PC, whatever...
One action on phone has same effect on server and vice versa.

Immich has only backup.
- You can add photos to your phone and it will backup to server. (this is called backup)
- You can add photos to server and they will not be on your phone. (this is not sync)
- You can use app to view photos on server (this is called browsing)
- If you don't have internet on your phone, you don't have photos.

There should be option to select which folders or photos you want to backup and which one you want to SYNC.

Folder One [✓]backup [✓]sync
Folder Two [✕]backup [✕]sync
Folder Three [✓]backup [✕]sync

If we had sync option, we could;
- have all data on the phone
- login to different server
- dump all photos, videos, folders, locations, tags and other metadata to new server

**Comment by mxlb-dev** [comment]
Please at least allow people to delete photos from apple library and then let immich delete them as well. My immich library is exploding and packed with images I don`t want or need and that have been deleted from my phone for a reason. I cannot believe I am the only one who takes multiple shots of the same thing in order to get a good one, and then delete the rest.

**Comment by chrneu** [comment]
I would agree it makes sense to start with the option that if a user deletes a photo on Apples Photo app, that this is synced to Immich. At least by having an option to respect deletions from the filesystem.

**Reply by mxlb-dev** [reply]
Yes, MVP could be the option to delete photos on immich server through deleting them on Apple/Google photos app + a flag to disable/enable that feature. Then everything else, two-way sync and all the other nice ideas in this thread and other discussions... :)

**Comment by B3N3-git** [comment]
When we delete a photo on the web, Immich could just ask for permission to delete the local file the next time we open the mobile app (similar to how Google Photos does it).
This way, there are no risky background deletions, the user stays in control, and we wouldn't have to delete everything twice.
Of course it should be possible to disable this feature for the users, who wants to keep all the photos in the native photos app

**Reply by PhilTdr** [reply]
This direction already exists (at least on Android as experimental feature) since [2.3.0](https://github.com/immich-app/immich/discussions/24007) introduced the “Delete synchronization” feature ([#3594](https://github.com/immich-app/immich/discussions/3594)). 

The other direction is still missing, as described in the previous comments. (Well, and maybe iOS 😇)

**Comment by glarue** [comment]
For anyone interested, as a workaround I built an iOS app (free, no ads, unaffiliated with Immich) for this from the client side called [`cairn`](https://github.com/glarue/cairn). Briefly, it watches `PHPhotoLibrary.fetchPersistentChanges` for `deletedLocalIdentifier`, joins on `AssetResponseDto.checksum`, holds confirmed deletions in a per-asset quarantine (14 days by default, configurable), then `DELETE /api/assets` with `force: false`. Nothing is permanently deleted; the assets get moved to the Immich server's Trash (at which point Immich's own trash policy takes over). As a breadcrumb, each run gets a `cairn/v1/run/<id>` tag you can use on the Immich side to audit a given `cairn` action. The code is all open source (MIT); app is iOS-only for now (it's what I currently use). Could easily be replaced by native behavior at some point in the future, but for now this was the shortest path-to-value for my own use. [iOS App Store link](https://apps.apple.com/app/id6763392945)

**Comment by naitkris** [comment]
I would like to suggest an optional age-based rule for two-way deletion sync on Android.

My wife and I use Immich on our Android phones, and we normally manage our recent photos in the phones’ native gallery apps. We often take several similar photos and later remove the duplicates, blurry photos, screenshots, or photos that simply did not turn out well. If we remove one of these recent photos, we genuinely want it removed from both the Android phone and the Immich server.

However, we also sometimes remove older photos because the storage on an Android phone is becoming full. In that situation, we only want to free up storage on the phone—the older photos should remain safely stored on Immich.

A possible solution could be an optional Android setting such as:

> Synchronize device deletions for photos newer than: **365 days**

With this enabled:

* If a photo newer than 365 days is removed through the Android phone’s native gallery, Immich could ask whether its server copy should also be moved to the Immich trash.
* If the photo is older than 365 days, removing it from the phone would not affect the server copy.
* Anything removed using Immich’s **Free Up Space** feature should always remain on the server, regardless of its age.

For example, I might take five similar photos today and delete four unwanted duplicates tomorrow using my normal Android gallery. Immich would detect this and ask if the same four photos should also be moved to the server trash.

Later, when I need more storage space, I might remove photos that are several years old. Those photos should disappear only from the Android phone and remain available on Immich.

For safety, this should be disabled by default and enabled separately on each Android device. The age limit could be configurable—perhaps 30, 90, 180, or 365 days—and server copies should first be moved to the Immich trash rather than permanently deleted.

Immich could also show a review message such as:

> 24 photos newer than 365 days were removed from this Android device. Would you also like to move their backed-up copies to the Immich trash?

If an unusually large number of photos suddenly disappears, Immich should pause and require confirmation. This would help protect against accidental deletion caused by changed storage permissions, a removed SD card, changed backup albums, reinstalling the app, or resetting the phone.

I think this would provide a useful middle ground between permanent one-way backup and unrestricted two-way synchronization: recent unwanted photos could be deleted everywhere, while older photos could remain safely archived on Immich when we remove them from our Android phones only to make more space.
